examples/rl/rllib/pg_pbt_example.py

```diff
@@ -0,0 +1,298 @@
+
+    print(f"======= Checkpointing at {str(result_dir)} =======")
+    pbt = PopulationBasedTraining(
+        time_attr="time_total_s",
+        metric="episode_reward_mean",
+        mode="max",
+        perturbation_interval=300,
+        resample_probability=0.25,
+        # Specifies the mutations of these hyperparams
+        # See: `ray.rllib.agents.trainer.COMMON_CONFIG` for common hyperparams
+        hyperparam_mutations={
+            "lr": [1e-3, 5e-4, 1e-4, 5e-5, 1e-5],
+            "rollout_fragment_length": lambda: rollout_fragment_length,
+            "train_batch_size": lambda: train_batch_size,
+        },
+        # Specifies additional mutations after hyperparam_mutations is applied
+        custom_explore_fn=explore,
+    )
+
+    from ray import air
+
+    run_config = air.RunConfig(
+        name=experiment_name,
+        stop={"time_total_s": time_total_s},
+        callbacks=[TrialCallback()],
+        storage_path=str(result_dir),
+        checkpoint_config=air.CheckpointConfig(
+            num_to_keep=3,
+            checkpoint_frequency=checkpoint_freq,
+            checkpoint_at_end=True,
+        ),
+        failure_config=air.FailureConfig(
+            max_failures=3,
+            fail_fast=False,
+        ),
+    )
+    tune_config = tune.TuneConfig(
+        num_samples=num_samples,
+        scheduler=pbt,
+        max_concurrent_trials=4,
+    )
+    trainable = "PG"
+    if resume_training:
+        tuner = tune.Tuner.restore(
+            str(experiment_dir),
+            trainable=trainable,
+            param_space=algo_config,
+        )
+    else:
+        tuner = tune.Tuner(
+            trainable=trainable,
+            param_space=algo_config,
+            tune_config=tune_config,
+            run_config=run_config,
+        )
+
+    results = tuner.fit()
+    # Get the best result based on a particular metric.
+    best_result = results.get_best_result(metric="episode_reward_mean", mode="max")
+
+    # Get the best checkpoint corresponding to the best result.
+    best_checkpoint = best_result.checkpoint
+
+    best_logdir = Path(best_result.log_dir)
+    model_path = best_logdir
+
+    copy_tree(str(model_path), save_model_path, overwrite=True)
+    print(f"Wrote model to: {save_model_path}")
+
+
+if __name__ == "__main__":
+    default_result_dir = str(
+        Path(__file__).resolve().parent / "results" / "tune_pg_results"
+    )
+    parser = gen_parser("rllib-example", default_result_dir)
+    parser.add_argument(
+        "--num_samples",
+        type=int,
+        default=1,
+        help="Number of times to sample from hyperparameter space",
+    )
+    parser.add_argument(
+        "--save_model_path",
+        type=str,
+        default=str(Path(__file__).expanduser().resolve().parent / "model"),
+        help="Destination path of where to copy the model when training is over",
+    )
+    args = parser.parse_args()
+    build_scenario(scenario=args.scenario, clean=False, seed=42)
```

Comment:
Here, using a seed, we first build the scenarios and generate a fixed traffic set. Then, we use the same scenarios in each of the the parallel environments to collect experience. Although the parallel environments might have different seeds and the scenario order might be shuffled, the underlying traffic set appears to be the same since they were generated using the same seed. Should we be building the scenarios with different seeds within each parallel environment instance?

Reply:
That approach will work only partially because unless we use a true-random seed, the seed is managed by `tune` in the `PBT` scheduling. From my investigation, all of the environments are identical within a trial. In that case, if we are building scenarios during the experiment, I am worried about race conditions between the environments because it is not clear how to differentiate between the environments.

The second issue is that the config is specified at the start of the configuration. To do what you ask would require copying and building the scenario at the start of each trial. I think that might be possible through the callbacks, but from what I can tell, the config is intended to be constant and I would need to do some investigation to see if it is possible.

What I am slightly surprised about ([see above comment](https://github.com/huawei-noah/SMARTS/pull/2067#issuecomment-1595354965)) is that the `config.vector_index` and `config.worker_index` values do not appear to be different across workers. And `reset(seed=None)` seems to be the constant case.

The easier way of working with this is to pre-generate a set of scenario variations at the start, I will expand the number of scenarios.